flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractHandler.java

```diff
@@ -221,7 +227,12 @@ private void finalizeRequestProcessing(FileUploads uploadedFiles) {
 
 	@Override
 	public final CompletableFuture<Void> closeAsync() {
-		return FutureUtils.composeAfterwards(closeHandlerAsync(), inFlightRequestTracker::awaitAsync);
+		if (isHandlerClosed.compareAndSet(false, true)) {
+			return FutureUtils.composeAfterwards(closeHandlerAsync(), inFlightRequestTracker::awaitAsync);
+		} else {
+			log.warn("The handler instance for {} had already been closed, but another attempt at closing it was made.", untypedResponseMessageHeaders.getTargetRestEndpointURL());
+			return CompletableFuture.completedFuture(null);
```

Comment:
I don't think we should return a completed future if the handle has received "close" message. It is possible that the handle is still closing. Possibly we alter ``inFlightRequestTracker::awaitAsync` to return the terminate future if it is closed, without deregister the phaser party.

cc @zentol  @GJL

Reply:
> I don't think we should return a completed future if the handle has received "close" message. It is possible that the handle is still closing. Possibly we alter `inFlightRequestTracker::awaitAsync` to return the terminate future if it is closed, without deregister the phaser party.
> 
> cc @zentol @GJL

Hi @TisonKun , thanks for your review. 

My original thought was that a completed ComplatableFuture *CompletableFuture.completedFuture(null)* would indicate that this invocation has finished without any longer effect, as it has nothing to do with the previous call.

To my understanding, if the REST handler was previously requested to be closed, then the corresponding CompletableFuture would have already been sent to the user, and if that CompletedFuture is still not finished, org.apache.flink.runtime.rest.RestServerEndpoint#closeHandlersAsync would wait for its completion, so it would not cause the RestServerEndpoint to shutdown earlier.

Or else maybe we could just throw an *Exception* if the handler is being closed more than once? This would make the design simpler.

Reply:
I am not sure whether we need to add such logics here. Since the `closeAsync` is not designed to be called multiple times and also currently we do not have this behavior. We close the handler one by one in `RestEndpoint#closeHandlersAsync`.

Reply:
> I am not sure whether we need to add such logics here. Since the `closeAsync` is not designed to be called multiple times and also currently we do not have this behavior. We close the handler one by one in `RestEndpoint#closeHandlersAsync`.

Hi @wangyang0918 , I believe that this code is to prevent future developer from accidentally reusing the same handlers multiple times. As you can see, the bug described in this JIRA ticket was caused by inappropriate reuse of one REST handler.

Reply:
@keijiyoshida I agree with you that the root cause is accidentally reusing the handler instance. However, we have checked the duplication in `RestServerEndpoint`, i am just think whether we still need to have the defensive logics here. 

I do not insist of this. You could determine by yourself.

Reply:
Hi @TisonKun , I have submitted a new commit which makes calls to closeAsync return the same result despite of multiple calls (idempotent).